src/parser/jobs/ast/modules/ArcanaTracking.tsx

```diff
@@ -0,0 +1,509 @@
+	}]
+	private pullStateInitialized = false
+	private pullIndex = 0
+
+	private prepullArcanas: BuffEvent[] = []
+
+	protected init() {
+		this.addHook('cast', {abilityId: CARD_ACTIONS, by: 'player'}, this.onCast)
+
+		this.addHook('applybuff', {abilityId: ARCANA_STATUSES, by: 'player'}, this.onPrepullArcana)
+		this.addHook('removebuff', {abilityId: ARCANA_STATUSES, by: 'player'}, this.offPrepullArcana)
+
+		this.addHook('removebuff', {abilityId: DRAWN_ARCANA, by: 'player'}, this.offDrawnStatus)
+		this.addHook('death', {to: 'player'}, this.onDeath)
+		this.addHook('complete', this.onComplete)
+	}
+
+	normalise(events: any) {
+		const startTime = this.parser.fight.start_time
+		let prepullSleeve = false
+		const sleeveDrawLog: CastEvent[] = []
+
+		for (const event of events) {
+			if ((event.timestamp > startTime && !prepullSleeve)
+				|| event.timestamp - startTime >= (STATUSES.SLEEVE_DRAW.duration * 1000)
+				) {
+				// End loop if: 1. Max duration of sleeve draw status passed 2. No prepull sleeve detected
+				break
+
+				} else if (!prepullSleeve && event.type === 'applybuff' && STATUSES.SLEEVE_DRAW.id === event.ability.guid
+				&& event.timestamp <= startTime) {
+					prepullSleeve = true
+				} else if (prepullSleeve && event.type === 'cast' && ACTIONS.DRAW.id === event.ability.guid) {
+					sleeveDrawLog.push(event)
+				} else if (prepullSleeve && event.type === 'removebuff' && STATUSES.SLEEVE_DRAW.id === event.ability.guid) {
+					prepullSleeve = false
+			} else {
+				continue
+			}
+		}
+
+		if (sleeveDrawLog.length > 0 && sleeveDrawLog.length <= 2) {
+			this.cardStateLog[0].sleeveState = sleeveDrawLog.length
+		}
+
+		return events
+
+	}
+
+	public get getCardLogs() {
+		if (this.cardStateLog.length === 0) {
+			return []
+		}
+		return this.cardStateLog
+	}
+
+	/**
+	 * @param {number} timestamp - desired timestamp to get the state. Defaults to pull state.
+	 * @returns {CardState} - object containing the card state and last event
+	 */
+	public getCardState(timestamp = this.parser.fight.start_time): CardState | undefined {
+		const stateItem = this.cardStateLog.find(artifact => artifact.lastEvent && timestamp > artifact.lastEvent.timestamp)
+		return stateItem
+	}
+
+	/**
+	 * Adds the Arcana Buff to _prepullArcanas if it was a precast status
+	 */
+	private onPrepullArcana(event: BuffEvent) {
+		if (event.timestamp !== this.parser.fight.start_time -1) {
+			return
+		}
+		this.prepullArcanas.push(event)
+	}
+
+	/**
+	 * Determine exactly when they casted their prepull arcana
+	 * If they had overwritten this buff, it will falsly pull back the timestamp of their prepull cast, but since we are guessing, it may as well be the same.
+	 */
+	private offPrepullArcana(event: BuffEvent) {
+		if (event.timestamp >= this.parser.fight.start_time + (STATUSES.THE_BALANCE.duration * 1000)) {
+			return
+		}
+
+		this.prepullArcanas.forEach(arcanaBuff => {
+			if (arcanaBuff.ability.guid === event.ability.guid
```

Comment:
if you invert this `if` and the one below it, you can `return` in them to shortcut out without needing to indent

Reply:
I need to master this technique :woah: